tools/asdf/src/proto.rs

```diff
@@ -0,0 +1,272 @@
+        return Err(PluginError::Message(String::from("Repository not found in downloaded file"))
+        .into());
+    };
+
+    Ok(Repo {
+        url: repo_url.to_string(),
+        default_branch: get_default_branch(&repo_url)?
+    })
+}
+
+
+// Workaround for a bug when using fs::remove_dir_all in this environment
+fn remove_dir_recursive(path: &VirtualPath) -> FnResult<()> {
+    let path = path.real_path().unwrap().into_os_string().into_string().unwrap();
+    if exec_command!("rm", ["-rf", &path]).exit_code != 0 {
+        return Err(PluginError::Message("Failed to remove directory".to_string()).into());
+    }
+    Ok(())
+}
+
+fn clone_repo(proto_temp_dir: VirtualPath) -> FnResult<VirtualPath> {
+    let repo = get_repo()?;
+    let repo_dir = virtual_path!(format!("{proto_temp_dir}/{}/repo", get_id(None)?));
+    // Remove the previous repo directory if it exists
+    remove_dir_recursive(&repo_dir)?;
+    fs::create_dir_all(&repo_dir.parent().unwrap())?;
+
+    if exec_command!("git", ["clone", "--depth=1", &repo.url, &repo_dir.real_path().unwrap().into_os_string().into_string().unwrap()]).exit_code != 0 {
+        return Err(PluginError::Message("Failed to clone repository".to_string()).into());
+    }
+
+    Ok(repo_dir)
+}
+
+fn get_versions(proto_temp_dir: VirtualPath) -> FnResult<Vec<String>> {
+    let script_path = clone_repo(proto_temp_dir)?;
+    let script_path = script_path.join("bin").join("list-all").real_path().unwrap().into_os_string().into_string().unwrap();
+
+    let versions = exec_command!("bash", [script_path]).stdout;
+    let versions: Vec<String> = versions.split_whitespace().map(str::to_owned).collect();
+    Ok(versions)
+}
+
+#[plugin_fn]
+pub fn register_tool(Json(input): Json<ToolMetadataInput>) -> FnResult<Json<ToolMetadataOutput>> {    
+    Ok(Json(ToolMetadataOutput {
+        name: input.id,
+        type_of: PluginType::Language,
+        minimum_proto_version: Some(Version::new(0, 45, 2)),
+        default_install_strategy: InstallStrategy::BuildFromSource,
+        plugin_version: Version::parse(env!("CARGO_PKG_VERSION")).ok(),
+		config_schema: Some(schematic::SchemaBuilder::generate::<AsdfPluginConfig>()),
+        ..ToolMetadataOutput::default()
+    }))
+}
+
+#[plugin_fn]
+pub fn build_instructions(
+    Json(input): Json<BuildInstructionsInput>,
+) -> FnResult<Json<BuildInstructionsOutput>> {
+    let env = get_host_environment()?;    
+
+    if env.os.is_windows() {
+        return Err(PluginError::UnsupportedWindowsBuild.into());
+    }
+    
+    let repo = get_repo()?;
+
+    let git = GitSource {
+        url: repo.url.clone(),
+        // Use default branch
+        reference: None,
+        ..Default::default()
+    };
+
+    let mut instructions = Vec::new();
+
+    let version = input.context.version;
+
+    // Set asdf environment variables
+    let install_download_path = &input.context.tool_dir.real_path().unwrap().into_os_string().into_string().unwrap();
```

Comment:
Would be safer to use `real_path!(&input.context.tool_dir)` or `real_path!(buf, ...)`. Can't remember which it is.